### PR TITLE
Convert a number of arrays to maps

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,17 +11,15 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
    Unreleased section, uncommenting the header as necessary.
 -->
 
-<!--## Unreleased-->
+## Unreleased
 
 * Added support for new JSDoc tags: @customElement, @polymer, @mixinFunction, @appliesMixin
+* [BREAKING] `attributes`, `properties`, `methods`, and `events` are now Maps from the name to the value rather than arrays. This better models what's actually going on (you can't have two different properties with the same name in jsavascript) and it makes our inheritance modelling more efficient.
```
Comment (web-padawan) [inline CHANGELOG.md]: typo: `jsavascript`

Comment (rictic) [inline CHANGELOG.md]: Good catch, thanks!

**Comment by rictic** [comment]
Reassigned to Fred and Al, as Justin is super busy in IO prep

File: src/javascript/esutil.ts [hunk 2]
```diff
@@ -126,14 +126,15 @@ export function getEventComments(node: estree.Node): ScannedEvent[] {
...
   });
-  return Array.from(eventComments)
-      .map(function(comment) {
-        const annotation =
-            jsdoc.parseJsdoc(jsdoc.removeLeadingAsterisks(comment).trim());
-        return annotateEvent(annotation);
-      })
-      .filter((ev) => !!ev)
-      .sort((ev1, ev2) => ev1.name.localeCompare(ev2.name));
+  const events = Array.from(eventComments)
```
Comment (usergenic) [inline src/javascript/esutil.ts]: arrow function golf
```typescript
const events = Array.from(eventComments)
  .map((comment) => annotateEvent(
      jsdoc.parseJsdoc(jsdoc.removeLeadingAsterisks(comment).trim()))
  .filter((ev) => !!ev)
  .sort((ev1, ev2) => ev1.name.localeCompare(ev2.name));
```

File: src/model/class.ts [hunk 3]
```diff
@@ -122,8 +121,8 @@ export class Class implements Feature {
   }
   readonly jsdoc: jsdocLib.Annotation|undefined;
   description: string;
-  readonly properties: Property[] = [];
-  readonly methods: Method[] = [];
+  readonly properties = new Map<string, Property>();
+  readonly methods: Map<string, Method> = new Map();
```
Comment (usergenic) [inline src/model/class.ts]: these two are defined differently? (methods with typedef on left and properties typedef is on the value) any particular reason?

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,7 +11,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
    Unreleased section, uncommenting the header as necessary.
 -->
 
-<!-- ## Unreleased -->
+## Unreleased
+
+* [BREAKING] `attributes`, `properties`, `methods`, and `events` are now Maps from the name to the value rather than arrays. This better models what's actually going on (you can't have two different properties with the same name in javascript) and it makes our inheritance modelling more efficient.
```
Comment (usergenic) [inline CHANGELOG.md]: Preferred spelling of modeling in the US is single L.  not exactly a deal-breaker.

Comment (rictic) [inline CHANGELOG.md]: oh wow yeah that does look weird too. fixed

File: src/javascript/esutil.ts (hunk 2 above)
Comment (rictic) [inline src/javascript/esutil.ts]: Works for me. Also changed the Array.from to use [...]

File: src/model/class.ts (hunk 3 above)
Comment (rictic) [inline src/model/class.ts]: no reason. standardized on putting the arguments on the constructor, as there are semi-rare cases where that does the right thing better

**Comment by rictic** [comment]
The more I use `[...x]` the more I like it. I'm in favor of standardizing on it, yeah.

**Pull request merged**